scanner/src/main/kotlin/experimental/ProvenanceBasedPostgresStorage.kt

```diff
@@ -0,0 +1,196 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import java.sql.SQLException
+
+import javax.sql.DataSource
+
+import org.jetbrains.exposed.dao.id.IntIdTable
+import org.jetbrains.exposed.sql.Column
+import org.jetbrains.exposed.sql.Database
+import org.jetbrains.exposed.sql.SchemaUtils
+import org.jetbrains.exposed.sql.SchemaUtils.withDataBaseLock
+import org.jetbrains.exposed.sql.and
+import org.jetbrains.exposed.sql.andWhere
+import org.jetbrains.exposed.sql.insert
+import org.jetbrains.exposed.sql.selectAll
+
+import org.ossreviewtoolkit.model.ArtifactProvenance
+import org.ossreviewtoolkit.model.KnownProvenance
+import org.ossreviewtoolkit.model.RepositoryProvenance
+import org.ossreviewtoolkit.model.ScanResult
+import org.ossreviewtoolkit.model.ScanSummary
+import org.ossreviewtoolkit.model.ScannerDetails
+import org.ossreviewtoolkit.model.utils.DatabaseUtils.checkDatabaseEncoding
+import org.ossreviewtoolkit.model.utils.DatabaseUtils.tableExists
+import org.ossreviewtoolkit.model.utils.DatabaseUtils.transaction
+import org.ossreviewtoolkit.scanner.ScannerCriteria
+import org.ossreviewtoolkit.scanner.storages.utils.jsonb
+import org.ossreviewtoolkit.utils.collectMessagesAsString
+import org.ossreviewtoolkit.utils.log
+import org.ossreviewtoolkit.utils.showStackTrace
+
+class ProvenanceBasedPostgresStorage(
+    /**
+     * The JDBC data source to obtain database connections.
+     */
+    private val dataSource: DataSource,
+
+    /**
+     * The name of the table used for storing scan results.
+     */
+    private val tableName: String = "provenance_scan_results"
+) : ProvenanceBasedScanStorage {
+    private val table = ProvenanceScanResults(tableName)
+
+    /** The [Database] instance on which all operations are executed. */
+    private val database = setupDatabase()
+
+    private fun setupDatabase(): Database =
+        Database.connect(dataSource).apply {
+            transaction {
+                withDataBaseLock {
+                    if (!tableExists(tableName)) {
+                        checkDatabaseEncoding()
+                        SchemaUtils.createMissingTablesAndColumns(table)
+                    }
+                }
+            }
+        }
+
+    override fun read(provenance: KnownProvenance): List<ScanResult> {
+        if (provenance is RepositoryProvenance && provenance.vcsInfo.path.isNotBlank()) {
```

Comment:
This check is duplicated at multiple places. Could it be extracted somewhere to a common helper function?